### PR TITLE
initial chrome API (navigation)

### PR DESCRIPTION
**Pull request by karelhala**
https://projects.engineering.redhat.com/browse/MAASPLAT-155

### Initial Chrome API
First version of Chrome Javascript API. The API is available under `insights.chrome` and exposes a few functions for manipulating sidenav. Also, the app can register listeners so that it can react to chrome navigation events.

Internally Redux is used for state keeping.

Very basic markup is used for now. This will be updated once https://projects.engineering.redhat.com/browse/MAASPLAT-167 is finished.

### Usage example
```js
    // initialize chrome
    insights.chrome.init();

    // identify yourself (the application). Currently this only tells Chrome which global navigation element to highlight
    insights.chrome.identifyApp('advisor');

    // define application navigation (navigation submenu)
    // at most one of the elements should be declared active
    // the operation is idempotent
    insights.chrome.navigation([{
        id: 'stability',
        title: 'Stability'
    }, {
        id: 'performance',
        title: 'Performance',
        active: true
    }]);

    // register a listener for application navigation events
    const unregister = insights.chrome.on('APP_NAVIGATION', event => {
        // change application route in response to navigation event from Chrome
        history.push(`/${event.navId}`);
    });

    // the listener can be unregistered if needed
    unregister();
```

### CODE REVIEW

**Comment by jharting** [comment]
Waiting for https://github.com/RedHatInsights/insights-frontend-components/pull/5

**Comment by Kinlaw** [comment]
@jharting I just merged RedHatInsights/insights-frontend-components#5

**Comment by jharting** [comment]
@Kinlaw awesome, thanks!

File: src/js/redux-config.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    from '@red-hat-insights/insights-frontend-components/Utilities/ReducerRegistry';
+import MiddlewareListener from '@red-hat-insights/insights-frontend-components/Utilities/MiddlewareListener';
+
+import * as appNav from './nav/appNav.js';
+import * as globalNav from './nav/globalNav.js';
+import { createReduxListener } from './utils';
+
+const basicMiddlewares = [];
+if (process.env.NODE_ENV === 'development') {
+    import('redux-logger').then(logger => basicMiddlewares.push(logger.default));
```
[the remark this answers is not in the document]
Comment (jharting) [reply src/js/redux-config.js]: Make sure you do `npm install`

**Comment by iphands** [comment]
@karelhala two questions... why does the client need to call insights.chrome.init?
Can you give an example of changing the active state?

**Comment by jharting** [comment]
@iphands 
> why does the client need to call insights.chrome.init?

Chrome needs to be initialized (redux store, initial state, etc.). `insights.chrome.init()` gives an application control over that process. That is useful because:
1. it helps with coordination and prevents timing issues. Since the app has control over Chrome's lifecycle it can be sure that Chrome will be initialized before the applications starts to use it
1. it allows the app to pass configuration options to Chrome that need to be defined before Chrome's initialization. There are no such options available currently but it's likely that we'll need a way for applications to configure Chrome especially if we go the "shared redux store" route (e.g. an app will want to pass reducers/middleware to chrome so that it's store is initialized with those)

> Can you give an example of changing the active state?

https://github.com/RedHatInsights/turbo-octo-couscous/pull/4/files

File: src/js/redux-config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kinlaw) [reply src/js/redux-config.js]: http://pastebin.test.redhat.com/601137

**Comment by Kinlaw** [comment]
Other than the test suite failing for me, this looks good.

**Comment by ryelo** [comment]
@jharting @karelhala , just want to point out that my test suite is failing too. Same messages as @Kinlaw

**Comment by jharting** [comment]
@ryelo @Kinlaw probably a Mac thing

**Comment by jharting** [comment]
@Kinlaw @ryelo can you `git pull && npm install && npm test` now?

**Comment by ryelo** [comment]
@jharting Passed for me!

**Comment by Kinlaw** [comment]
works for me! merging

**Pull request merged**